README.md
# geo-tools

[![Build Status](https://app.travis-ci.com/lykmapipo/geo-tools.svg?branch=master)](https://app.travis-ci.com/lykmapipo/geo-tools)
[![Dependencies Status](https://david-dm.org/lykmapipo/geo-tools.svg)](https://david-dm.org/lykmapipo/geo-tools)
[![Coverage Status](https://coveralls.io/repos/github/lykmapipo/geo-tools/badge.svg?branch=master)](https://coveralls.io/github/lykmapipo/geo-tools?branch=master)
[![GitHub License](https://img.shields.io/github/license/lykmapipo/geo-tools)](https://github.com/lykmapipo/geo-tools/blob/master/LICENSE)

[![Commitizen Friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg)](https://github.com/prettier/prettier)
[![Code Style](https://badgen.net/badge/code%20style/airbnb/ff5a5f?icon=airbnb)](https://github.com/airbnb/javascript)
[![npm version](https://img.shields.io/npm/v/@lykmapipo/geo-tools)](https://www.npmjs.com/package/@lykmapipo/geo-tools)

Opinionated geo utilities for day to day development

## Requirements

- [NodeJS v8.11.1+](https://nodejs.org)
- [npm v5.6.0+](https://www.npmjs.com/)

## Installation

```sh
npm install --save @lykmapipo/geo-tools
```

## Usage

```js
import {
  randomPosition,
  randomPoint,
  readShapefile,
  readGeoJSON
} from '@lykmapipo/geo-tools';

randomPosition()
//=> [-76.41031, 67.0704]

randomPoint()
//=> { type: 'Point', coordinates: [ -76.41031, 67.0704 ] }

readShapefile('data.shp', (error, { finished, feature, next }) => {
  // handle read errors
  if (error) {
    return done(error);
  } 
  // handle read finish
  if (finished) {
    return done();
  } 
  // process features
  // and request next chunk from stream
  processFeature(error => next(error));
});

readGeoJSON('data.geojson', (error, { finished, feature, next }) => {
  // handle read errors
  if (error) {
    return done(error);
  } 
  // handle read finish
  if (finished) {
    return done();
  } 
  // process features
  // and request next chunk from stream
  processFeature(error => next(error));
});
```

## Environment
```js
GEO_MAX_LENGTH=0.0001
GEO_MAX_ROTATION=0.39269908169872414
GEO_BBOX=-180,-90,180,90
```

## Test

- Clone this repository

- Install all dependencies

```sh
npm install
```

- Then run test

```sh
npm test
```

## Contribute

It will be nice, if you open an issue first so that we can know what is going on, then, fork this repo and push in your ideas. Do not forget to add a bit of test(s) of what value you adding.

## Licence

The MIT License (MIT)

Copyright (c) lykmapipo & Contributors

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
#### 0.7.27 (2021-09-06)

##### Chores

* **deps:**  force latest version & audit fix ([6504548c](https://github.com/lykmapipo/common/commit/6504548c7db59ebd4e9d2728add1b550ee73488c))

##### Documentation Changes

* **readme:**  update build status to travis.com ([11dd5ec6](https://github.com/lykmapipo/common/commit/11dd5ec6d73483ba35b5a1d6265da3a706e1c9cc))

#### 0.7.26 (2021-08-09)

##### Chores

* **deps:**  force latest version & audit fix ([d21f8854](https://github.com/lykmapipo/common/commit/d21f88549af0e788166f436986d8b68759faf407))

##### Documentation Changes

* **readme:**  update status badges ([da1e4d62](https://github.com/lykmapipo/common/commit/da1e4d6203847ca58479fbed7fe8544ca62f575e))
*  add security guide ([3d5d65e2](https://github.com/lykmapipo/common/commit/3d5d65e2e67db6542da0d5ba9cbcc708289ba425))

#### 0.7.25 (2021-06-04)

##### Chores

* **deps:**  force latest version & audit fix ([1dc9825e](https://github.com/lykmapipo/common/commit/1dc9825ea5aba7cd6c04d2ac35468dd89bcfe63e))

#### 0.7.24 (2021-02-09)

##### Chores

* **deps:**  force latest version & audit fix ([34426268](https://github.com/lykmapipo/common/commit/34426268c4a6a4ed1f71c1ec9db91804e4def0be))

#### 0.7.23 (2021-01-04)

##### Chores

* **deps:**  force latest version & audit fix ([3a8aa5a7](https://github.com/lykmapipo/common/commit/3a8aa5a7e492b4db760d98b2c23d6c71791c3d3b))

#### 0.7.22 (2020-12-25)

#### 0.7.21 (2020-12-16)

##### Chores

* **deps:**  force latest version & audit fix ([d2163fc1](https://github.com/lykmapipo/common/commit/d2163fc135c5b9d3f10a4bb9c2b66eb39f727f24))

#### 0.7.20 (2020-12-05)

##### Chores

* **deps:**  force latest version & audit fix ([33a8e288](https://github.com/lykmapipo/common/commit/33a8e2887608127e0cd604103bbd8063b4a33ea9))

#### 0.7.19 (2020-11-12)

#### 0.7.18 (2020-10-05)

##### Chores

* **deps:**  force latest version & audit fix ([20a4e442](https://github.com/lykmapipo/common/commit/20a4e442373d570348c59e7e74174b1dc83d93c6))

#### 0.7.17 (2020-09-23)

#### 0.7.16 (2020-09-10)

##### Chores

* **deps:**  force latest version & audit fix ([0119bcca](https://github.com/lykmapipo/common/commit/0119bccad4a8adcedfcd7c18feeb65b1a41265a9))

#### 0.7.15 (2020-09-02)

##### Chores

* **deps:**  force latest version & audit fix ([9934b0b4](https://github.com/lykmapipo/common/commit/9934b0b423b202ba63f1ec42befa22c1d7747b54))

#### 0.7.14 (2020-08-19)

##### Chores

* **deps:**  force latest version & audit fix ([0e1a6f61](https://github.com/lykmapipo/common/commit/0e1a6f61725b02767993c0365bae349f84bbccea))

#### 0.7.13 (2020-08-04)

##### Chores

* **deps:**  force latest version & audit fix ([859760ce](https://github.com/lykmapipo/common/commit/859760ce0af0e81f5b7bf82c47c6a47432e8de3e))

#### 0.7.12 (2020-07-27)

##### Chores

* **deps:**  force latest version & audit fix ([5f89e193](https://github.com/lykmapipo/common/commit/5f89e193d1babc6398db959ac0435005cdd1568b))

#### 0.7.11 (2020-07-20)

##### Chores

* **deps:**  force latest version & audit fix ([e902ecaa](https://github.com/lykmapipo/common/commit/e902ecaa50c66683a1d1aedc55262332263e133c))

#### 0.7.10 (2020-07-12)

##### Chores

* **deps:**  force latest version & audit fix ([1fd2af50](https://github.com/lykmapipo/common/commit/1fd2af50408dc28847278255052e1655fea9d0c5))

#### 0.7.9 (2020-07-08)

##### Chores

* **deps:**  force latest version & audit fix ([018ce951](https://github.com/lykmapipo/common/commit/018ce951689ef59d800920cfaab14d5071022bad))

#### 0.7.8 (2020-07-01)

##### Chores

* **deps:**  force latest version & audit fix ([f49a6b68](https://github.com/lykmapipo/common/commit/f49a6b6872d683df7a7bd927d20c82e6b4eb9f49))

#### 0.7.7 (2020-06-25)

##### Chores

* **deps:**  force latest version & audit fix ([e8e0a9d6](https://github.com/lykmapipo/common/commit/e8e0a9d68befa4976aa1f58f8143acdfe3ff221f))

#### 0.7.6 (2020-06-16)

##### Chores

* **deps:**  force latest version & audit fix ([324710de](https://github.com/lykmapipo/common/commit/324710decb6c54a4a749147bd9c22483802c11b0))

#### 0.7.5 (2020-06-07)

##### Chores

* **deps:**  force latest version & audit fix ([835427fe](https://github.com/lykmapipo/common/commit/835427fed8f28eb5cf3ac858a9db673e73cb0891))

#### 0.7.4 (2020-05-29)

##### Chores

* **deps:**  force latest version & audit fix ([87ae2578](https://github.com/lykmapipo/common/commit/87ae2578a3bbbc9227ac1b1b366094afedf6d6c7))

#### 0.7.3 (2020-05-14)

##### Chores

* **deps:**  force latest version & audit fix ([2b5134dd](https://github.com/lykmapipo/common/commit/2b5134dde287fb8ea4b0a631bc5aa4c8702cc097))

#### 0.7.2 (2020-05-09)

##### Chores

* **deps:**  force latest version & audit fix ([3f01c152](https://github.com/lykmapipo/common/commit/3f01c152bccc44e1e6f6fe0c9773292e52fc5003))

#### 0.7.1 (2020-05-01)

##### Chores

* **deps:**  force latest version & audit fix ([ffb532c3](https://github.com/lykmapipo/common/commit/ffb532c31a8c22030ebf755552e23213a89f4698))

#### 0.7.0 (2020-05-01)

##### Chores

* **deps:**  force latest version & audit fix ([f510c6f4](https://github.com/lykmapipo/common/commit/f510c6f4554ccb10024648e9803e7705f34e15ba))

##### Refactors

*  use newer geojson-validation api ([70dc98a2](https://github.com/lykmapipo/common/commit/70dc98a2041897a9f820af3b55f75d7c3882ddde))

#### 0.6.9 (2020-04-25)

##### Chores

* **deps:**  force latest version & audit fix ([f68bb891](https://github.com/lykmapipo/common/commit/f68bb8910f73b912cf764c592b6be241f670c8aa))

#### 0.6.8 (2020-04-15)

##### Chores

* **deps:**  force latest version & audit fix ([6af0a2f6](https://github.com/lykmapipo/common/commit/6af0a2f6847261d6077bc20d5c58440e63307cb2))

#### 0.6.7 (2020-03-30)

##### Chores

* **deps:**  force latest version & audit fix ([ad6113a1](https://github.com/lykmapipo/common/commit/ad6113a1e389f50377dd82dfd513152180d1e029))

#### 0.6.6 (2020-03-06)

##### Chores


src/random.js
import { forEach, map, range, sample, sampleSize } from 'lodash';
import { mergeObjects } from '@lykmapipo/common';
import {
  GEO_MAX_LENGTH,
  GEO_MAX_ROTATION,
  GEO_BBOX,
  GEO_POINT,
  GEO_LINESTRING,
  GEO_POLYGON,
  GEO_MULTIPOINT,
  GEO_MULTILINESTRING,
  GEO_MULTIPOLYGON,
  GEO_GEOMETRY_COLLECTION,
} from './constant';

/**
 * @function randomLongitude
 * @name randomLongitude
 * @description Generate random longitude
 * @param {object} [optns={}] valid option
 * @param {number[]} [optns.bbox=[-180, -90, 180, 90]] a bounding box inside
 * of which geometries are placed.
 * @returns {object} valid longitude
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 * const longitude = randomLongitude();
 * // => -76.4103176657406
 */
export const randomLongitude = (optns = {}) => {
  const { bbox = GEO_BBOX } = optns;
  const longitude = Math.random() * (bbox[2] - bbox[0]) + bbox[0];
  return longitude;
};

/**
 * @function randomLatitude
 * @name randomLatitude
 * @description Generate random latitude
 * @param {object} [optns={}] valid option
 * @param {number[]} [optns.bbox=[-180, -90, 180, 90]] a bounding box inside
 * of which geometries are placed.
 * @returns {object} valid latitude
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 * const latitude = randomLatitude();
 * // => 67.07040223216296
 */
export const randomLatitude = (optns = {}) => {
  const { bbox = GEO_BBOX } = optns;
  const latitude = Math.random() * (bbox[3] - bbox[1]) + bbox[1];
  return latitude;
};

/**
 * @function randomPosition
 * @name randomPosition
 * @description Generate next random position
 * @param {object} [optns={}] valid option
 * @param {number} [optns.angle = (Math.random() * 2 * Math.PI)] valid angle in radian between points
 * @param {number} [optns.distance = (Math.random() * 0.0001)] valid distance between points
 * @param {number} [optns.longitude] valid longitude on last point
 * @param {number} [optns.latitude] valid latitude on last point
 * @param {number[]} [optns.bbox=[-180, -90, 180, 90]] a bounding box inside
 * of which geometries are placed.
 * @returns {object} valid position
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 * const position = randomPosition();
 * // => [ -76.4103176657406, 67.07040223216296 ]
 */
export const randomPosition = (optns = {}) => {
  // TODO: restric within bbox

  // calculate angle
  const angle =
    (optns.angle || Math.random() * 2 * Math.PI) +
    (Math.random() - 0.5) * GEO_MAX_ROTATION * 2;

  // calculate hypotenus
  const distance = optns.distance || Math.random() * GEO_MAX_LENGTH;

  // x2 = x1 + dcos0
  const x1 = optns.longitude || randomLongitude(optns);
  const x2 = x1 + distance * Math.cos(angle);

  // y2 = y1 + dsin0
  const y1 = optns.latitude || randomLatitude(optns);
  const y2 = y1 + distance * Math.sin(angle);

  // return x2,y2
  return [x2, y2];
};

/**
 * @function randomPositions
 * @name randomPositions
 * @description Generate random positions
 * @param {object} [optns={}] valid option
 * @param {number} [optns.vertices=2] how many positions.
 * @param {number[]} [optns.bbox=[-180, -90, 180, 90]] a bounding box inside
 * of which geometries are placed.
 * @returns {object} valid positions
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example const geo = randomPositions();
 * // => [ [-76.41031, 67.0704], ...]
 */
export const randomPositions = (optns = { vertices: 2 }) => {
  // ensure options
  const options = mergeObjects({ vertices: 2, bbox: GEO_BBOX }, optns);

  // refs
  let coordinates = [];
  let longitude;
  let latitude;

  // compute position for vertices
  forEach(range(options.vertices), () => {
    // random next position after last
    const position = randomPosition(
      mergeObjects(options, { longitude, latitude })
    );
    [longitude, latitude] = position;
    // collect position
    coordinates = [...coordinates, position];
  });

  // return positions
  return coordinates;
};

/**
 * @function randomPoint
 * @name randomPoint
 * @description Generate random GeoJSON Point
 * @param {object} [optns={}] valid option
 * @param {number[]} [optns.bbox=[-180, -90, 180, 90]] a bounding box inside
 * of which geometries are placed.
 * @returns {object} valid GeoJSON Point
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example const geo = randomPoint();
 * // => { type: 'Point', coordinates:[ ... ] }
 */
export const randomPoint = (optns = {}) => {
  // refs
  const type = GEO_POINT;
  const coordinates = randomPosition(optns);

  // return point
  return { type, coordinates };
};

/**
 * @function randomLineString
 * @name randomLineString
 * @description Generate random GeoJSON LineString
 * @param {object} [optns={}] valid option
 * @param {number} [optns.vertices=2] how many coordinates each LineString will contain.
 * @param {number[]} [optns.bbox=[-180, -90, 180, 90]] a bounding box inside
 * of which geometries are placed.
 * @returns {object} valid GeoJSON LineString
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example const geo = randomLineString();
 * // => { type: 'LineString', coordinates:[ ... ] }
 */
export const randomLineString = (optns = { vertices: 2 }) => {
  // refs
  const type = GEO_LINESTRING;
  const coordinates = randomPositions(optns);

  // return linestring
  return { type, coordinates };
};

/**
 * @function randomPolygon
 * @name randomPolygon
 * @description Generate random GeoJSON Polygon
 * @param {object} [optns={}] valid option
 * @param {number} [optns.vertices=4] how many coordinates Polygon will contain.
 * @param {number[]} [optns.bbox=[-180, -90, 180, 90]] a bounding box inside
 * of which geometries are placed.
 * @returns {object} valid GeoJSON Polygon
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const geo = randomPolygon();
 * // => { type: 'Polygon', coordinates:[ ... ] }
 */
export const randomPolygon = (optns = { vertices: 3 }) => {
  // ensure 4 vertices & above
  const options = mergeObjects({ vertices: 3 }, optns);
  options.vertices = options.vertices < 3 ? 3 : options.vertices;

  // refs
  const type = GEO_POLYGON;
  let coordinates = randomPositions(options);

  // close the ring
  coordinates = [...coordinates, coordinates[0]];

  // return linestring
  return { type, coordinates: [coordinates] };
};

/**
 * @function randomMultiPoint
 * @name randomMultiPoint
 * @description Generate random GeoJSON MultiPoint
 * @param {object} [optns={}] valid option
 * @param {number} [optns.vertices=2] how many points MultiPoint will contain.
 * @param {number[]} [optns.bbox=[-180, -90, 180, 90]] a bounding box inside
 * of which geometries are placed.
 * @returns {object} valid GeoJSON MultiPoint
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const geo = randomMultiPoint();
 * // => { type: 'MultiPoint', coordinates:[ ... ] }
 */
export const randomMultiPoint = (optns = { vertices: 2 }) => {
  // refs
  const type = GEO_MULTIPOINT;
  const coordinates = randomPositions(optns);

  // return multipoint
  return { type, coordinates };
};

/**
 * @function randomMultiLineString
 * @name randomMultiLineString
 * @description Generate random GeoJSON MultiLineString
 * @param {object} [optns={}] valid option
 * @param {number} [optns.lines=2] how many LineString.
 * @param {number} [optns.vertices=2] how many coordinates each LineString will contain.
 * @param {number[]} [optns.bbox=[-180, -90, 180, 90]] a bounding box inside
 * of which geometries are placed.
 * @returns {object} valid GeoJSON MultiLineString
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const geo = randomMultiLineString();
 * // => { type: 'MultiLineString', coordinates:[ ... ] }
 */
export const randomMultiLineString = (optns = { lines: 2, vertices: 2 }) => {
  // ensure lines & vertices
  const options = mergeObjects({ lines: 2, vertices: 2 }, optns);

  // refs
  const type = GEO_MULTILINESTRING;
  const coordinates = [
    ...map(range(options.lines), () => {
      return randomPositions(options);
    }),
  ];

  // return multilinestring
  return { type, coordinates };
};

/**
 * @function randomMultiPolygon
 * @name randomMultiPolygon
 * @description Generate random GeoJSON MultiPolygon
 * @param {object} [optns={}] valid option
 * @param {number} [optns.polygons=2] how many Polygons.
 * @param {number} [optns.vertices=2] how many coordinates each Polygon will contain.
 * @param {number[]} [optns.bbox=[-180, -90, 180, 90]] a bounding box inside
 * of which geometries are placed.
 * @returns {object} valid GeoJSON MultiPolygon
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const geo = randomMultiPolygon();
 * // => {type: 'MultiPolygon', coordinates:[] }
 */
export const randomMultiPolygon = (optns = { polygons: 2, vertices: 3 }) => {
  // ensure polygons & vertices
  const options = mergeObjects({ polygons: 2, vertices: 3 }, optns);

  // refs
  const type = GEO_MULTIPOLYGON;
  const coordinates = [
    ...map(range(options.polygons), () => {
      const coords = randomPositions(options);
      return [[...coords, coords[0]]];
    }),
  ];

  // return multipolygon
  return { type, coordinates };
};

/**
 * @function randomGeometry
 * @name randomGeometry
 * @description Generate random GeoJSON Geometry
 * @param {object} [optns={}] valid option
 * @returns {object} valid GeoJSON Geometry
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const geo = randomGeometry();
 * // => { type: 'Point', coordinates:[ ... ] }
 */
export const randomGeometry = (optns = {}) => {
  // ensure options
  const options = mergeObjects({ lines: 2, polygons: 2, vertices: 3 }, optns);

  // geometry generators
  const generators = [
    randomPoint,
    randomLineString,
    randomPolygon,
    randomMultiPoint,
    randomMultiLineString,
    randomMultiPolygon,
  ];

  // generate geometry
  const generateGeomentry = sample(generators);
  const geometry = generateGeomentry(options);

  // retunr geometry
  return geometry;
};

/**
 * @function randomGeometryCollection
 * @name randomGeometryCollection
 * @description Generate random GeoJSON GeometryCollection
 * @param {object} [optns={}] valid option
 * @returns {object} valid GeoJSON GeometryCollection
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since 0.1.0
 * @version 0.1.0
 * @static
 * @public
 * @example
 *
 * const geo = randomGeometryCollection();
 * // => { type: 'GeometryCollection', geometries: [ ... ] }
 */
export const randomGeometryCollection = (optns = {}) => {
  // ensure options
  const options = mergeObjects({ lines: 2, polygons: 2, vertices: 3 }, optns);

  // geometry generators
  const generators = [
    randomPoint,
    randomLineString,
    randomPolygon,
    randomMultiPoint,
    randomMultiLineString,
    randomMultiPolygon,
  ];

  // refs
  const type = GEO_GEOMETRY_COLLECTION;

  // generate geometry
  const geometries = map(
    sampleSize(generators, options.vertices),
    (generateGeomentry) => {
      return generateGeomentry(options);
    }
  );

  // return geometrycollection
  return { type, geometries };
};
